File: minicluster/src/main/java/alluxio/multi/process/Worker.java
/*
 * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
 * (the "License"). You may not use this work except in compliance with the License, which is
 * available at www.apache.org/licenses/LICENSE-2.0
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied, as more fully set forth in the License.
 *
 * See the NOTICE file distributed with this work for information regarding copyright ownership.
 */

package alluxio.multi.process;

import alluxio.conf.PropertyKey;

import com.google.common.base.Preconditions;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Closeable;
import java.io.File;
import java.io.IOException;
import java.util.Map;

import javax.annotation.concurrent.ThreadSafe;

/**
 * Class for running and interacting with an Alluxio worker in a separate process.
 */
@ThreadSafe
public class Worker implements Closeable {
  private static final Logger LOG = LoggerFactory.getLogger(Worker.class);

  private final File mLogsDir;
  private final Map<PropertyKey, String> mProperties;

  private ExternalProcess mProcess;

  /**
   * @param logsDir log directory
   * @param properties alluxio properties
   */
  public Worker(File logsDir, Map<PropertyKey, String> properties) throws IOException {
    mLogsDir = logsDir;
    mProperties = properties;
  }

  /**
   * Launches the worker process.
   */
  public synchronized void start() throws IOException {
    Preconditions.checkState(mProcess == null, "Worker is already running");
    mProcess = new ExternalProcess(mProperties, LimitedLifeWorkerProcess.class,
        new File(mLogsDir, "worker.out"));
    mProcess.start();
  }

  @Override
  public synchronized void close() {
    if (mProcess != null) {
      mProcess.stop();
      mProcess = null;
    }
  }
}

File: cli/src/alluxio.org/cli/processes/worker.go
/*
 * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
 * (the "License"). You may not use this work except in compliance with the License, which is
 * available at www.apache.org/licenses/LICENSE-2.0
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied, as more fully set forth in the License.
 *
 * See the NOTICE file distributed with this work for information regarding copyright ownership.
 */

package processes

import (
	"fmt"
	"strings"

	"github.com/palantir/stacktrace"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"

	"alluxio.org/cli/env"
)

var Worker = &WorkerProcess{
	BaseProcess: &env.BaseProcess{
		Name:                 "worker",
		JavaClassName:        "alluxio.worker.AlluxioWorker",
		JavaOpts:             ConfAlluxioWorkerJavaOpts,
		ProcessOutFile:       "worker.out",
		MonitorJavaClassName: "alluxio.worker.AlluxioWorkerMonitor",
	},
}

const (
	envAlluxioWorkerLogger = "ALLUXIO_WORKER_LOGGER"
	workerLoggerType       = "WORKER_LOGGER"
)

var (
	ConfAlluxioWorkerJavaOpts = env.RegisterTemplateEnvVar(&env.AlluxioConfigEnvVar{
		EnvVar:     "ALLUXIO_WORKER_JAVA_OPTS",
		IsJavaOpts: true,
	})
	confAlluxioWorkerAttachOpts = env.RegisterTemplateEnvVar(&env.AlluxioConfigEnvVar{
		EnvVar:     "ALLUXIO_WORKER_ATTACH_OPTS",
		IsJavaOpts: true,
	})
)

type WorkerProcess struct {
	*env.BaseProcess
}

func (p *WorkerProcess) Base() *env.BaseProcess {
	return p.BaseProcess
}

func (p *WorkerProcess) SetEnvVars(envVar *viper.Viper) {
	envVar.SetDefault(envAlluxioWorkerLogger, workerLoggerType)
	// ALLUXIO_WORKER_JAVA_OPTS = {default logger opts} ${ALLUXIO_JAVA_OPTS} ${ALLUXIO_WORKER_JAVA_OPTS}
	javaOpts := []string{
		fmt.Sprintf(env.JavaOptFormat, env.ConfAlluxioLoggerType, envVar.Get(envAlluxioWorkerLogger)),
	}
	javaOpts = append(javaOpts, env.ConfAlluxioJavaOpts.JavaOptsToArgs(envVar)...)
	javaOpts = append(javaOpts, p.JavaOpts.JavaOptsToArgs(envVar)...)
	envVar.Set(p.JavaOpts.EnvVar, strings.Join(javaOpts, " "))
}

func (p *WorkerProcess) StartCmd(cmd *cobra.Command) *cobra.Command {
	cmd.Use = p.Name
	return cmd
}

func (p *WorkerProcess) startCmdArgs() []string {
	cmdArgs := []string{env.Env.EnvVar.GetString(env.ConfJava.EnvVar)}
	cmdArgs = append(cmdArgs, confAlluxioWorkerAttachOpts.JavaOptsToArgs(env.Env.EnvVar)...)
	cmdArgs = append(cmdArgs, "-cp", env.Env.EnvVar.GetString(env.EnvAlluxioServerClasspath))
	cmdArgs = append(cmdArgs, p.JavaOpts.JavaOptsToArgs(env.Env.EnvVar)...)

	// specify a default of -Xmx4g if no memory setting is specified
	const xmxOpt = "-Xmx"
	if !argsContainsOpt(cmdArgs, xmxOpt, "MaxRAMPercentage") {
		cmdArgs = append(cmdArgs, fmt.Sprintf("%v4g", xmxOpt))
	}
	// specify a default of -XX:MaxDirectMemorySize=4g if not set
	const maxDirectMemorySize = "-XX:MaxDirectMemorySize"
	if !argsContainsOpt(cmdArgs, maxDirectMemorySize) {
		cmdArgs = append(cmdArgs, fmt.Sprintf("%v=4g", maxDirectMemorySize))
	}
	return append(cmdArgs, p.JavaClassName)
}

func (p *WorkerProcess) Start(cmd *env.StartProcessCommand) error {
	if err := p.Launch(cmd, p.startCmdArgs()); err != nil {
		return stacktrace.Propagate(err, "error launching process")
	}
	return nil
}

func (p *WorkerProcess) StopCmd(cmd *cobra.Command) *cobra.Command {
	cmd.Use = p.Name
	return cmd
}
